apps/enterprise-knowledge-retrieval/enterprise_knowledge_retrieval.ipynb

```diff
@@ -0,0 +1,1750 @@
+   "metadata": {},
+   "source": [
+    "### Evaluation\n",
+    "\n",
+    "Last comes the not-so-fun bit that will make the difference between nifty prototype and production application - the process of evaluating and tuning your results. \n",
+    "\n",
+    "The key takeaway here is to make a framework that saves the results of each evaluation, as well as the parameters. Evaluation can be a difficult task that takes significant resources, so it is best to start prepared to handle multiple iterations. Some useful principles we've seen successful deployments use are:\n",
+    "- **Assign clear product ownership and metrics:** Ensure you have a team aligned from the start to annotate the outputs and determine whether they're bad or good. This may seem an obvious step, but too often the focus is on the engineering challenge of successfully retrieving content rather than the product challenge of providing retrieval results that are useful.\n",
+    "- **Log everything:** Store all requests and responses to and from your LLM and retrieval service if you can, it builds a great base for fine-tuning both the embeddings and any fine-tuned models or few-shot LLMs in future.\n",
+    "- **Use GPT-4 as a labeller:** When running evaluations, it can help to use GPT-4 as a gatekeeper for human annotation. Human annotation is costly and time-consuming, so doing an initial evaluation run with GPT-4 can help set a quality bar that needs to be met to justify human labeling. At this stage I would not suggest using GPT-4 as your only labeler, but it can certainly ease the burden.\n",
+    "    - This approach is outlined further in [this paper](https://arxiv.org/abs/2108.13487).\n",
+    "\n",
+    "We'll use these principles to make a quick evaluation framework where we will:\n",
+    "- Use GPT-4 to make a list of hypothetical questions on our topic\n",
+    "- Ask our Agent the questions and save question/answer tuples\n",
+    "    - These two above steps simulate the actual users interacting with your application\n",
+    "- Get GPT-4 to evaluate whether the answers correctly respond to the questions\n",
+    "- Look at our results to measure how well the Agent answered the questions\n",
+    "- Plan remedial action"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 83,
+   "id": "b8314f7b",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import time\n",
+    "\n",
+    "# Build a prompt to provide the original query, the result and ask to summarise for the user\n",
+    "evaluation_question_prompt = '''You are a helpful Wikipedia assistant who generates creative general knowledge questions.\n",
+    "\n",
+    "Examples:\n",
+    "- Explain how photons work\n",
+    "- What is Thomas Dolby known for?\n",
+    "- What are some key events of the 20th century?\n",
+    "\n",
+    "Begin!\n",
+    "\n",
+    "Question:'''\n",
+    "\n",
+    "evaluation_questions = []\n",
+    "\n",
+    "for i in range(0,10):\n",
```

Comment:
Instead of looping sequentially 10 times, you can also use the `n` API parameter to get 10 completions in a single go.

Reply:
On second thought, I wouldn't suggest either approach. I noticed a lot of duplicate or near-duplicate questions in the list. I think it might make more sense to ask for a single list of 100 questions than a bunch of short lists using an identical prompt, because even with high temperature there are a lot of repeats.

Reply:
Went with this updated approach of 10 questions.